From 05a06b8a498cc9558299fcad89fa4d0d40da9cad Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Mon, 26 Aug 2013 15:37:50 +0700
Subject: Dalvik: fillVectorizationInformation handles constant incorrectly

BZ: 133399

fillVectorizationInformation check whether the def is constant
even when def is 0. The patch fixes that.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: Ia737ee3d9c8b620f3a55a5734ec7530c09087b64
Orig-MCG-Change-Id: I57a3cda601189fff1751b4a4cb30ef7ef8b6c3be
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
Reviewed-on: http://android.intel.com:8080/127683
Reviewed-by: Petrochenko, Dmitry <dmitry.petrochenko@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Reviewed-by: Koski, Anttu <anttu.koski@intel.com>
Reviewed-by: Beyler, Jean Christophe <jean.christophe.beyler@intel.com>
Reviewed-by: Lupusoru, Razvan A <razvan.a.lupusoru@intel.com>
Reviewed-by: Processe, Christian <christian.processe@intel.com>
Tested-by: Processe, Christian <christian.processe@intel.com>
Reviewed-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Reviewed-by: Kraiem, RihabX <rihabx.kraiem@intel.com>
Tested-by: Kraiem, RihabX <rihabx.kraiem@intel.com>
---
 vm/compiler/Vectorization.cpp |   10 +++-------
 1 files changed, 3 insertions(+), 7 deletions(-)

diff --git a/vm/compiler/Vectorization.cpp b/vm/compiler/Vectorization.cpp
index e19efb2..5ba71c7 100644
--- a/vm/compiler/Vectorization.cpp
+++ b/vm/compiler/Vectorization.cpp
@@ -575,13 +575,9 @@ static bool fillVectorizationInformation (const CompilationUnit *cUnit, LoopInfo
             }
             else
             {
-                setOutputRegister (info, vrUsed, true);
-            }
-
-            //Finally, if the define is a CONST bytecode, remove it as an output, and remove the constant
-            if ((dvmCompilerDataFlowAttributes[defMir->dalvikInsn.opcode] & DF_SETS_CONST) != 0 )
-            {
-                setOutputRegister (info, vrUsed, false);
+                //Finally, if the define is a CONST bytecode, remove it as an output, and remove the constant
+                bool isNotAConst = (dvmCompilerDataFlowAttributes[defMir->dalvikInsn.opcode] & DF_SETS_CONST) == 0;
+                setOutputRegister (info, vrUsed, isNotAConst);
             }
 
             //Go to the next use
-- 
1.7.4.1

